coderd/notifications/reports/generator.go

```diff
@@ -0,0 +1,296 @@
+		}
+
+		templateAdmins, err := findTemplateAdmins(ctx, db, stats)
+		if err != nil {
+			logger.Error(ctx, "unable to find template admins", slog.F("template_id", stats.TemplateID), slog.Error(err))
+			continue
+		}
+
+		for _, templateAdmin := range templateAdmins {
+			reportLog, err := db.GetReportGeneratorLogByUserAndTemplate(ctx, database.GetReportGeneratorLogByUserAndTemplateParams{
+				UserID:                 templateAdmin.ID,
+				NotificationTemplateID: notifications.TemplateWorkspaceBuildsFailedReport,
+			})
+			if err != nil && !xerrors.Is(err, sql.ErrNoRows) { // sql.ErrNoRows: report not generated yet
+				logger.Error(ctx, "unable to get recent report generator log for user", slog.F("user_id", templateAdmin.ID), slog.Error(err))
+				continue
+			}
+
+			if !reportLog.LastGeneratedAt.IsZero() && reportLog.LastGeneratedAt.Add(failedWorkspaceBuildsReportFrequencyDays*24*time.Hour).After(now) {
+				// report generated recently, no need to send it now
+				continue
+			}
+
+			reportGeneratedNow[templateAdmin.ID] = true
+
+			if len(failedBuilds) == 0 {
+				// no failed workspace builds, no need to send the report
+				continue
+			}
+
+			templateDisplayName := stats.TemplateDisplayName
+			if templateDisplayName == "" {
+				templateDisplayName = stats.TemplateName
+			}
+
+			if _, err := enqueuer.EnqueueData(ctx, templateAdmin.ID, notifications.TemplateWorkspaceBuildsFailedReport,
+				map[string]string{
+					"template_name":         stats.TemplateName,
+					"template_display_name": templateDisplayName,
+				},
+				reportData,
+				"report_generator",
+				stats.TemplateID, stats.TemplateOrganizationID,
+			); err != nil {
+				logger.Warn(ctx, "failed to send a report with failed workspace builds", slog.Error(err))
+			}
+		}
+	}
+
+	for recipient := range reportGeneratedNow {
+		err = db.UpsertReportGeneratorLog(ctx, database.UpsertReportGeneratorLogParams{
+			UserID:                 recipient,
+			NotificationTemplateID: notifications.TemplateWorkspaceBuildsFailedReport,
+			LastGeneratedAt:        dbtime.Time(now).UTC(),
+		})
+		if err != nil {
+			logger.Error(ctx, "unable to update report generator logs", slog.F("user_id", recipient), slog.Error(err))
+		}
+	}
+
+	err = db.DeleteOldReportGeneratorLogs(ctx, database.DeleteOldReportGeneratorLogsParams{
+		NotificationTemplateID: notifications.TemplateWorkspaceBuildsFailedReport,
+		Before:                 dbtime.Time(now.Add(-failedWorkspaceBuildsReportFrequencyDays*24*time.Hour - time.Hour)).UTC(),
+	})
+	if err != nil {
+		return xerrors.Errorf("unable to delete old report generator logs: %w", err)
+	}
+	return nil
+}
+
+func buildDataForReportFailedWorkspaceBuilds(frequencyDays int, stats database.GetWorkspaceBuildStatsByTemplatesRow, failedBuilds []database.GetFailedWorkspaceBuildsByTemplateIDRow) map[string]any {
+	// Format frequency label
+	var frequencyLabel string
+	if frequencyDays == 7 {
+		frequencyLabel = "week"
```

Comment:
This seems like the only possibility at this time, why do we need this handling?

Reply:
In the original design there was an idea to make it customizable, but right now I can definitely simplify it. Thanks for spotting this!